### PR TITLE
Geometry_Engine: NurbsSurfaces: Degrees() fix and IsClosed, IsPeriodic implementation

### LINKED ISSUE #1297: Geometry_Engine: Remove Degrees() method for NurbsSurface
#### Description:
`BH.Engine.Geometry.Query.Degrees(this NurbsSurface surf)` method does not return correct results when the first knot value does not have full multiplicity ([Rhino wiki](https://wiki.mcneel.com/rhino/nurbs)). Therefore, a new approach is proposed in [this BHoM PR](https://github.com/BHoM/BHoM/pull/603) and #1298.

#### Test file:
Uploaded to [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FBHoM%5FEngine%2FGeometry%5FEngine%2FGeometry%5FEngine%2DIssue1297%2DRemoveDegreesMethodForNurbsSurface).

### LINKED ISSUE #1320: Geometry_Engine: add methods to check whether a nurbs surface is periodic and closed
#### Description:
Since some software packages (e.g. Revit) have issues with periodic and closed nurbs surfaces, it would be good to have a method that would check it before pushing.

### PR DESCRIPTION
**Pull request by pawelbaran**
### NOTE: Depends on 
[#603](https://github.com/BHoM/BHoM/pull/603)

   
### Issues addressed by this PR

Closes #1297 
Closes #1320

### Test files

Test files are located on [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FBHoM%2FGeometry%5FoM%2DIssue425%2DCreationOfTrimmedNurbsSurfaceDefinition). To run it, please switch to the branches related to the following PRs:
- [#603 ](https://github.com/BHoM/BHoM/pull/603)
- [#1319](https://github.com/BHoM/BHoM_Engine/pull/1319)
- [#122](https://github.com/BHoM/Rhinoceros_Toolkit/pull/122)
- [#435](https://github.com/BHoM/Grasshopper_Toolkit/pull/435)
- [#435](https://github.com/BHoM/Revit_Toolkit/pull/435)
- [#183](https://github.com/BHoM/Dynamo_Toolkit/pull/183)

The error in Rhino -> BHoM -> Rhino convert error in the test file is caused by the bug in Cone convert logged [here](https://github.com/BHoM/Rhinoceros_Toolkit/issues/121).

### Changelog

- `BH.Engine.Geometry.Query.Degrees(this NurbsSurface surf)` has been removed.
- `BH.Engine.Geometry.Query.IsClosed(this NurbsSurface surface)` has been added.
- `BH.Engine.Geometry.Query.IsPeriodic(this NurbsSurface surface)` has been added.

### Additional comments
Updated copy of #1298.
Please see [#603](https://github.com/BHoM/BHoM/pull/603).

Issues to be raised after merging this:
- implement `InternalEdges(NurbsSurface surface)` and `ExternalEdges(NurbsSurface surface)` in `BH.Engine.Geometry.Query`

### CODE REVIEW

File: Geometry_Engine/Query/Degree.cs [hunk 1]
```diff
@@ -37,18 +37,5 @@ public static int Degree(this NurbsCurve curve)
         }
 
         /***************************************************/
-
-        public static List<int> Degrees(this NurbsSurface surf)
```
Comment (epignatelli) [inline Geometry_Engine/Query/Degree.cs]: Can we leave this method, deprecate it and make it return the `UDegree` and `VDegree` of the input `NurbsSurface`?

Nothing wrong about removing it, it's just to give some buffer to people that might have used it.
We can then remove it in the next beta

File: Geometry_Engine/Query/IsPeriodic.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace BH.Engine.Geometry
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static bool IsPeriodic(this NurbsSurface surface)
```
Comment (epignatelli) [inline Geometry_Engine/Query/IsPeriodic.cs]: Did you figure out the logic at the end?

File: Geometry_Engine/Query/UVCount.cs [hunk 3]
```diff
@@ -33,8 +33,7 @@ public static partial class Query
 
         public static List<int> UVCount(this NurbsSurface surf)
         {
-            List<int> degrees = surf.Degrees();
```
Comment (epignatelli) [inline Geometry_Engine/Query/UVCount.cs]: If you keep the `Degrees` method, you can avoid changing this as well.

File: Geometry_Engine/Query/IsPeriodic.cs (hunk 2 above)
Comment (pawelbaran) [reply Geometry_Engine/Query/IsPeriodic.cs]: I hope so 🙈

File: Geometry_Engine/Query/Degree.cs (hunk 1 above)
Comment (pawelbaran) [reply Geometry_Engine/Query/Degree.cs]: You are right - actually I am not allowed to remove the method (can only deprecate it). I will do the latter and do my best to remove all references to it.

File: Geometry_Engine/Query/UVCount.cs (hunk 3 above)
Comment (pawelbaran) [reply Geometry_Engine/Query/UVCount.cs]: I will implement the change in order to avoid using a deprecated method.

**Comment by epignatelli** [comment]
/azp why are you giving us a red tick?

@FraserGreenroyd any idea?

**Comment by FraserGreenroyd** [comment]
@epignatelli see https://github.com/BHoM/Test_Toolkit/issues/68 - trying to look into this this morning :smile:

**Comment by FraserGreenroyd** [comment]
> /azp why are you giving us a red tick?
> 
> @FraserGreenroyd any idea?

Should be fixed now @epignatelli :smile:

**Pull request merged**